AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,130 @@
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+    if (!tokenArray) {
+      tokenArray = [NSMutableArray<MSAuthTokenInfo *> new];
+    }
+    if (tokenArray.count == 0) {
+
+      // Add nil token if the entire story is empty.
+      [tokenArray addObject:[MSAuthTokenInfo new]];
+    }
+    if (![tokenArray.lastObject.authToken isEqual:authToken]) {
+
+      // If new token differs from the last token of array - add it to array.
+      MSAuthTokenInfo *newAuthToken = [[MSAuthTokenInfo alloc] initWithAuthToken:authToken andStartTime:[NSDate date] andEndTime:expiresOn];
+      [tokenArray addObject:newAuthToken];
+    }
+
+    // Cap array size at max available size const (deleting from beginning).
+    if ([tokenArray count] > kMSIdentityMaxAuthTokenArraySize) {
+      [tokenArray removeObjectAtIndex:0];
+    }
+
+    // Save new array.
+    if ([MSKeychainUtil storeArray:tokenArray forKey:kMSIdentityAuthTokenArrayKey]) {
+      MSLogDebug([MSIdentity logTag], @"Saved new history state in keychain.");
+    } else {
+      MSLogWarning([MSIdentity logTag], @"Failed to save new history state in keychain.");
+    }
+    if (authToken) {
+      [self saveTokenToKeychain:authToken];
+    } else {
+      [self deleteTokenFromKeychain];
+    }
+    if (authToken && accountId) {
+      [MS_USER_DEFAULTS setObject:(NSString *)accountId forKey:kMSIdentityMSALAccountHomeAccountKey];
+    } else {
+      [MS_USER_DEFAULTS removeObjectForKey:kMSIdentityMSALAccountHomeAccountKey];
+    }
+  }
+}
+
+- (void)removeAuthToken:(nullable NSString *)authToken {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+
+    // Do nothing if there's just one entry in the history or no history at all.
+    if (!tokenArray || tokenArray.count == 1) {
+      return;
+    }
+
+    // Find, delete the oldest entry. Do not delete the most recent entry.
+    for (NSUInteger i = 0; i < tokenArray.count - 1; i++) {
+      if ([tokenArray[i] authToken] == authToken) {
+        [tokenArray removeObjectAtIndex:i];
+        break;
+      }
+    }
+
+    // Save new array after changes.
+    [MSKeychainUtil storeArray:tokenArray forKey:kMSIdentityAuthTokenArrayKey];
+  }
+}
+
+- (void)deleteTokenFromKeychain {
+  if ([MSKeychainUtil deleteStringForKey:kMSIdentityAuthTokenKey]) {
+    MSLogDebug([MSIdentity logTag], @"Deleted auth token from keychain.");
+  } else {
+    MSLogWarning([MSIdentity logTag], @"Failed to delete auth token from keychain.");
+  };
+}
+
+- (void)saveTokenToKeychain:(nullable NSString *)authToken {
+  if ([MSKeychainUtil storeString:(NSString *)authToken forKey:kMSIdentityAuthTokenKey]) {
```

Comment:
Remove `kMSIdentityAuthTokenKey` and use the last token from history

Reply:
Removed, thanks. There are 9 tests failing after this change. We will fix them and remove the unused constant a bit later.